[WIP] Ner pipeline grouped_entities fixes

cceyda opened the pull request:
There are many issues with ner pipeline using grouped_entities=True
https://github.com/huggingface/transformers/issues/5077
https://github.com/huggingface/transformers/issues/4816
https://github.com/huggingface/transformers/issues/5730
https://github.com/huggingface/transformers/issues/5609
https://github.com/huggingface/transformers/issues/6514
https://github.com/huggingface/transformers/issues/5541

- [x]  [Bug Fix] add an option `ignore_subwords` to ignore subsequent ##wordpieces in predictions. Because some models train on only the first token of a word and not on the subsequent wordpieces (BERT NER default). So it makes sense doing the same thing at inference time.

    - The simplest fix is to just group the subwords with the first wordpiece. 
      - [TODO] how to handle ignored scores? just set them to 0 and calculate zero invariant mean ?
      - [TODO] handle different wordpiece_prefix ## ? possible approaches:
get it from tokenizer? but currently most tokenizers dont have a wordpiece_prefix property?
have an _is_subword(token)

- [x] [Feature add] added option to `skip_special_tokens`. Cause It was harder to remove them after grouping.

- [x] [Additional Changes] remove B/I prefix on returned grouped_entities 

Edit: Ignored subwords' scores are also ignored by setting them to nan and using nanmean
Edit: B entities of different type are separated (as per BIO tag definition)
Edit: skip_special_tokens is now the default behavior
Edit: ignore_subwords is now the default behavior
Edit: more flexibility for custom non-standard tokenizers through tokenizer.is_subword_fn, tokenizer.convert_tokens_to_string
Edit: [fix UNK token related bugs by mapping UNK tokens to the correct original string] Use fast tokenizer or pass offset_mapping

# Usage
`pipeline('ner', model=model, tokenizer=tokenizer, ignore_labels=[], grouped_entities=True, ignore_subwords=True)`

cceyda commented on src/transformers/pipelines.py:
```diff
@@ -1112,10 +1126,17 @@ def group_entities(self, entities: List[dict]) -> List[dict]:
+            # Shouldn't merge if both entities are B-type
             if (
                 entity["entity"].split("-")[-1] == entity_group_disagg[-1]["entity"].split("-")[-1]
+                and entity["entity"].split("-")[0] != "B"
```
`B-` is supposed to signify the start of a new entity by the definition of [IOB2 format](https://en.wikipedia.org/wiki/Inside%E2%80%93outside%E2%80%93beginning_(tagging)), which is commonly used for example by CONLL ner task.
(Ah this was supposed to be in a separate [PR](https://github.com/huggingface/transformers/pull/5966)...but I guess I miss-commited)

enzoampil replied:
I understand. I think this is reasonable, I'm just thinking of potential cases where an entity is labelled `B-` at the middle of a token. For example, in this recent [issue](https://github.com/huggingface/transformers/issues/4816) where `B-` is used for two consecutive word pieces. For these cases, I think the word pieces should still be grouped.

Perhaps we can account for this added complexity in a separate PR from this one then? I'll suggest to remove this line first.

```
[
{'word': 'Cons', 'score': 0.9994944930076599, 'entity': 'B-PER', 'index': 1},
{'word': '##uelo', 'score': 0.802545428276062, 'entity': 'B-PER', 'index': 2}
]
```

enzoampil commented on src/transformers/pipelines.py:
```diff
@@ -1112,10 +1126,17 @@ def group_entities(self, entities: List[dict]) -> List[dict]:
 
             # If the current entity is similar and adjacent to the previous entity, append it to the disaggregated entity group
             # The split is meant to account for the "B" and "I" suffixes
+            # Shouldn't merge if both entities are B-type
```
```suggestion
```

enzoampil commented on src/transformers/pipelines.py:
```diff
@@ -1112,10 +1126,17 @@ def group_entities(self, entities: List[dict]) -> List[dict]:
+            # Shouldn't merge if both entities are B-type
             if (
                 entity["entity"].split("-")[-1] == entity_group_disagg[-1]["entity"].split("-")[-1]
+                and entity["entity"].split("-")[0] != "B"
```
```suggestion
```

cceyda replied:
Now I'm thinking maybe this should be optional?

enzoampil replied:
@cceyda Actually, I reviewed again, and agree we should keep this here.

HHoofs commented:
I'm wondering should the B & I part maybe separated from the entity type part? In the sense that you average the entities (disregarding the B/I part) and vice-versa. I now have the feeling that only the first subtoken decides whether the complete word is a B or an I.

cceyda commented:
I want to complete this but ran into another issue while working it:

All [UNK] tokens get mapped to [UNK] in the output, instead of the actual input token (because the code is getting from ids->tokens), Also [UNK]s gets lost when using skip_special_tokens (https://github.com/huggingface/transformers/issues/6863)
While this is a simple token alignment issue and can be solved by using `offset_mappings`. `offset_mappings` is only available with fast tokenizers, I'm wondering what would be a more general approach to solving this?

Monique497 commented:
Dear @cceyda,

In the last couple of days I started to work with transformers and especially NER-classification. I ran into issues that has been previously addressed in other issues you just mentioned at the beginning. Especially that subtokens that were classified with 'O'  were not properly merged with the full token. 

For example (Dutch):
sentence = "Als we Volkswagens OR-voorzitter **Bernd Osterloh** moeten geloven, dan moet dat binnen drie jaar het geval zijn."

Gives me as group entities: 
[{'entity_group': 'B-per', 'score': 0.9999980926513672, 'word': 'Bern'}, 
{'entity_group': 'I-per', 'score': 0.9999990463256836, 'word': 'Ost'}]

I expect:
[{'entity_group': 'B-per', 'score': 0.9999980926513672, 'word': 'Bernd'}, 
{'entity_group': 'I-per', 'score': 0.9999990463256836, 'word': 'Osterloh'}]

However, the considered subtokens are classified as 'O':

{'word': '[CLS]', 'score': 0.9999999403953552, 'entity': 'O', 'index': 0}
{'word': 'Als', 'score': 0.9999999403953552, 'entity': 'O', 'index': 1}
{'word': 'we', 'score': 0.9999999403953552, 'entity': 'O', 'index': 2}
{'word': 'Volkswagen', 'score': 0.9999955296516418, 'entity': 'B-misc', 'index': 3}
{'word': '##s', 'score': 0.9999999403953552, 'entity': 'O', 'index': 4}
{'word': 'O', 'score': 0.9981945157051086, 'entity': 'I-misc', 'index': 5}
{'word': '##R', 'score': 0.9999998807907104, 'entity': 'O', 'index': 6}
{'word': '-', 'score': 0.9999999403953552, 'entity': 'O', 'index': 7}
{'word': 'voorzitter', 'score': 0.9999998807907104, 'entity': 'O', 'index': 8}
{'word': 'Bern', 'score': 0.9999980926513672, 'entity': 'B-per', 'index': 9}
**{'word': '##d', 'score': 0.9999998807907104, 'entity': 'O', 'index': 10}**
{'word': 'Ost', 'score': 0.9999990463256836, 'entity': 'I-per', 'index': 11}
**{'word': '##er', 'score': 0.9999998807907104, 'entity': 'O', 'index': 12}
{'word': '##lo', 'score': 0.9999997615814209, 'entity': 'O', 'index': 13}
{'word': '##h', 'score': 0.9999998807907104, 'entity': 'O', 'index': 14}**
{'word': 'moeten', 'score': 0.9999999403953552, 'entity': 'O', 'index': 15}
{'word': 'geloven', 'score': 0.9999998807907104, 'entity': 'O', 'index': 16}
{'word': ',', 'score': 0.9999999403953552, 'entity': 'O', 'index': 17}
{'word': 'dan', 'score': 0.9999999403953552, 'entity': 'O', 'index': 18}
{'word': 'moet', 'score': 0.9999999403953552, 'entity': 'O', 'index': 19}
{'word': 'dat', 'score': 0.9999999403953552, 'entity': 'O', 'index': 20}
{'word': 'binnen', 'score': 0.9999999403953552, 'entity': 'O', 'index': 21}
{'word': 'drie', 'score': 0.9999999403953552, 'entity': 'O', 'index': 22}
{'word': 'jaar', 'score': 0.9999999403953552, 'entity': 'O', 'index': 23}
{'word': 'het', 'score': 0.9999999403953552, 'entity': 'O', 'index': 24}
{'word': 'geval', 'score': 0.9999999403953552, 'entity': 'O', 'index': 25}
{'word': 'zijn', 'score': 0.9999999403953552, 'entity': 'O', 'index': 26}
{'word': '.', 'score': 0.9999999403953552, 'entity': 'O', 'index': 27}
{'word': '[SEP]', 'score': 0.9999999403953552, 'entity': 'O', 'index': 28}

I believe your pull request addresses these issues properly. 
However, I saw the merge did not complete since it failed on some tasks.

I was wondering if there is still the intention to solve these issues.

Disclaimer: I am a total newbie to git (just set up an account), so please be mild, haha.
Any help is much appreciated!

Thank you in advance,

Monique

enzoampil commented:
@cceyda I actually want this PR to move forward. Are you okay collaborating on your fork (can add me as collaborator)? I can help out with some of the issues failing so we can get this merged :smile:

cceyda commented:
@enzoampil I have added you as a collaborator.
Also pushed some additional changes addressing the [UNK] token mapping problem I mentioned before. 
Still there are some things I'm not very satisfied with:

1. subword prefix was fixed to '##' before. with the latest change I added a check to see if the tokenizer has an `is_subword_fn `defined (still dont like handling it this way). I know some tokenizers have `subword_prefix` but most don't and this was the most flexible solution for now. 
2. `offset_mappings`  is needed to resolve [UNK] tokens, but is only available with fast tokenizers. Fast tokenizers don't have `convert_ids_to_tokens` so had to implement a hacky solution for those aswell.
3.  `skip_special_tokens` also dropped [UNK] tokens so I had to change things and rely on `special_tokens_mask`.

It is not optimal but it worked for my use cases. 
Haven't had a chance to look at the failing tests yet :/

cceyda commented:
I have changed the `ignore_subwords` default to True which covers cases like
```
[
{'word': 'Cons', 'score': 0.9994944930076599, 'entity': 'B-PER', 'index': 1},
{'word': '##uelo', 'score': 0.802545428276062, 'entity': 'B-PER', 'index': 2}
]
```
And honestly I don't know why subwords shouldn't be ignored for most cases. (Unless there is need for some custom logic that determines a words tag; ie by averaging the wordpieces etc etc. In which case grouped_entities shouldn't be used 🤔  )
IMO Mid-word inconsistencies made by the model while `ignore_subwords = False` shouldn't effect pipelines output logic.

[todo]
- torch tests are passing for now but probably should add more cases? (I can't see why the tf tests are failing though, don't have dev env for that)
- should add the new parameters to the doc strings.

Monique497 commented:
This seems good, now, doesn't it? :) Thank you for picking this up, again. Me and my team really appreciate that.

Monique497 commented:
Dear @cceyda,

Last two days I worked on your branch to see how it performs on my own input texts. 
However, I  came accross the following issue I would like to point out to you:

When I use the following line of code (as you suggest under 'Usage' above):

pipeline('ner', model=model, tokenizer=tokenizer, ignore_labels=[], grouped_entities=True, skip_special_tokens=True, ignore_subwords=True)


I get the error:

TypeError: __init__() got an unexpected keyword argument 'skip_special_tokens'.

When looking in the file transformer.pipelines and looking specifically for the tokenclassificationpipeline, it seems that it is not yet implemented. Or am I missing something?

Best, 

Monique

cceyda commented:
@Monique497 sorry for the delay
A couple of things have changed since I first wrote that example:
 
- special tokens ([CLS][PAD][SEP]) are always skipped (per comments above) so you don't need that kwarg.

``` py
from transformers import (
    AutoModelForTokenClassification,
    AutoTokenizer,
    pipeline,
)

model = AutoModelForTokenClassification.from_pretrained(model_name)
tokenizer = AutoTokenizer.from_pretrained(model_name, use_fast=True) # note the fast tokenizer use
# ignore_subwords = True by default
nlp = pipeline("ner",model=model,tokenizer=tokenizer, grouped_entities=True)
inputs="test sentence"
output=nlp(inputs)
```

- Another important thing is you have to **use a fast tokenizer** OR pass `offset_mapping` as a parameter because the [UNK] token resolution depends on this. (maybe I should rename this to offset_mappings)

```py
# you can pass it like this
nlp(inputs,offset_mapping=mappings_you_calculate)

```
- If you are using a custom tokenizer that treats subwords differently (ie not starting with '##'), you can pass a function implementing your custom logic through `tokenizer.is_subword_fn`  and `tokenizer.convert_tokens_to_string` 
I don't know if this is the best way to handle non standard tokenizations, but I use some custom non-standard tokenizers for Korean and this solution gave me enough flexibility.

something like this:

```py
def sub_fn(token):
    if token.starts_with("%%"): return True
tokenizer.is_subword_fn=sub_fn

def convert_tokens_to_string(self, tokens):
    out_string = " ".join(tokens).replace(" %%", "").strip()
    return out_string
tokenizer.convert_tokens_to_string=convert_tokens_to_string
```

@enzoampil what are your thoughts on this?

cceyda commented on tests/test_pipelines.py:
```diff
@@ -771,8 +785,15 @@ def _test_ner_pipeline(
 
-        for ungrouped_input, grouped_result in zip(ungrouped_ner_inputs, expected_grouped_ner_results):
-            self.assertEqual(nlp.group_entities(ungrouped_input), grouped_result)
+        if nlp.grouped_entities:
```
conditioned so that grouped_entities=False tests won't fail because of grouped_entities=True

enzoampil replied:
LGTM

cceyda commented on tests/test_pipelines.py:
```diff
@@ -771,8 +785,15 @@ def _test_ner_pipeline(
-        for ungrouped_input, grouped_result in zip(ungrouped_ner_inputs, expected_grouped_ner_results):
-            self.assertEqual(nlp.group_entities(ungrouped_input), grouped_result)
+        if nlp.grouped_entities:
+            if nlp.ignore_subwords:
```
added case for ignore_subwords=True and False

cceyda reviewed (commented):
There were some more changes after this, if you can take a look at the latest.

enzoampil commented:
@cceyda Sorry for taking a while, lemme do another review!

enzoampil reviewed (approved):
Awesome work @cceyda ! I think this looks good to go :smile:

enzoampil commented:
@LysandreJik @julien-c This looks good to me. Please advise if this is good to merge or if you think there's still anything missing before merging :grin:

LysandreJik commented on src/transformers/pipelines.py:
```diff
@@ -1382,7 +1387,14 @@ def __call__(self, *args, **kwargs):
+                    return_offsets_mapping=self.tokenizer.is_fast,
                 )
+                if "offset_mapping" in tokens:
+                    offset_mapping = tokens["offset_mapping"].cpu().numpy()[0]
+                    del tokens["offset_mapping"]
+                special_tokens_mask = tokens["special_tokens_mask"].cpu().numpy()[0]
```
This seems like it would break if ran from a tensorflow pipeline

cceyda replied:
oh I see that, will fix

LysandreJik replied:
Alright, thanks for making sure it works!

LysandreJik commented on src/transformers/pipelines.py:
```diff
@@ -1435,17 +1457,31 @@ def group_sub_entities(self, entities: List[dict]) -> dict:
         }
         return entity_group
 
+    def is_subword_fn(self, token: str) -> bool:
+        if token.startswith("##"):
+            return True
+        return False
```
This is very specific to BERT

cceyda replied:
I provide alternatives for those BERT specific functions; the user can override the function by providing their own tokenizer.is_subword_fn and tokenizer.convert_tokens_to_string. while it is not a clean solution it is flexible. I would like it if there was a property like tokenizer.subword_prefix for all tokenizers, then I could use that. (I remember I saw something similar somewhere maybe in the fast tokenizers?)

LysandreJik replied:
I'm pretty sure for slow tokenizers one can use `tokenizer.convert_tokens_to_string`, while for fast tokenizer you can use `token_to_word` to obtain the indices and construct words.

I don't think expecting the user to provide a method to join tokens is the best we can do here. Pipelines are meant as a way for users to use the models without needing to understand what happens under the hood, having them specify a tokenizer-specific method to join tokens goes against that goal.

cceyda replied:
I agree about the pipelines, hadn't liked my previous hacky solution myself either. I have changed tokenizer.convert_tokens_to_string bit as suggested https://github.com/huggingface/transformers/pull/5970#discussion_r511790482. and also the subword detection logic.https://github.com/huggingface/transformers/pull/5970#discussion_r511789502

LysandreJik commented on src/transformers/pipelines.py:
```diff
@@ -1435,17 +1457,31 @@ def group_sub_entities(self, entities: List[dict]) -> dict:
+            return True
+        return False
+
+    def convert_tokens_to_string(self, tokens):
+        """ Converts a sequence of tokens (string) in a single string. """
+        if hasattr(self.tokenizer, "convert_tokens_to_string"):
+            # fast tokenizers dont have convert_tokens_to_string?!
+            return self.tokenizer.convert_tokens_to_string(tokens)
+        else:
+            out_string = " ".join(tokens).replace(" ##", "").strip()
+            return out_string
```
This too is very specific to BERT

cceyda replied:
same here. can be overridden by tokenizer.convert_tokens_to_string. I recently saw the [BatchEncoding.token_to_word](https://huggingface.co/transformers/main_classes/tokenizer.html#transformers.BatchEncoding.token_to_word) maybe I can use that, will take a look

LysandreJik replied:
Can we use the `token_to_word` if it's available, `convert_tokens_to_string` otherwise?

LysandreJik reviewed (changes requested):
Thanks for iterating on this! There seems to be an issue with the TF tests, as it doesn't seem to be testing the pipeline in itself (see comment).

Also, this approach uses very BERT-specific splitting/joining, which we can't have as we should aim to support all 30 models.

jgibson commented on tests/test_pipelines.py:
```diff
@@ -736,15 +736,29 @@ def _test_ner_pipeline(
-                {"entity_group": "B-PER", "score": 0.9997273534536362, "word": "Andrés Pastrana"},
-                {"entity_group": "B-ORG", "score": 0.8589080572128296, "word": "Farc"},
+                {"entity_group": "PER", "score": 0.999369223912557, "word": "Consuelo Araújo Noguera"},
+                {"entity_group": "PER", "score": 0.9997771680355072, "word": "Andrés Pastrana"},
```
A test for hyphenated names (ex. Juantia Gomez-Cortez) would be useful, especially given that the fast and slow tokenizers have different codepaths for reconstructing the original text. I had to implement grouping of named entities myself recently and was tripped up by that corner case.

cceyda commented on src/transformers/pipelines.py:
```diff
@@ -1426,15 +1452,14 @@ def __call__(self, *args, **kwargs):
+                start_ind, end_ind = offset_mapping[idx]
+                word_ref = sentence[start_ind:end_ind]
+                word = self.tokenizer.convert_ids_to_tokens([int(input_ids[idx])])[0]
+                is_subword = len(word_ref) != len(word)
```
changed `is_subword` detection logic: by comparing length of token(##token) with the original text span mapping (Assuming subwordpieces get prefixed by something). 
Incase the user wants some other logic they can first get ungrouped entities add `is_subword:bool` field to entities and call `pipeline.group_entities` themselves.

cceyda commented on src/transformers/pipelines.py:
```diff
@@ -1467,28 +1495,17 @@ def group_sub_entities(self, entities: List[dict]) -> dict:
+        if self.tokenizer.is_fast:
+            word = self.tokenizer.decode(self.tokenizer.convert_tokens_to_ids(tokens))
+        else:
+            word = self.tokenizer.convert_tokens_to_string(tokens)
```
fixed as suggested! I agree it is much cleaner this way.

cceyda commented on src/transformers/pipelines.py:
```diff
@@ -1299,6 +1299,29 @@ def __call__(self, *args, targets=None, **kwargs):
         return results
 
 
+class TokenClassificationArgumentHandler(ArgumentHandler):
```
added this to check `offset_mapping` if provided. (does a simple batch_size check)

cceyda commented on tests/test_pipelines_ner.py:
```diff
@@ -105,53 +107,66 @@ def _test_pipeline(self, nlp: Pipeline):
-        for model_name in self.small_models:
-            nlp = pipeline(
-                task="ner", model=model_name, tokenizer=model_name, grouped_entities=True, ignore_subwords=False
+                task="ner", model=model_name, tokenizer=tokenizer, grouped_entities=True, ignore_subwords=True
```
Don't know why the tests are failing, I get normal results when running them outside of the test suite 😕 ?

LysandreJik commented:
Thanks for iterating! I'll check this today.

LysandreJik reviewed (approved):
Played around with it, works well, and the implementation seems robust. Looks good, LGTM! Thanks for iterating.

@stefan-it, do you want to take a quick look?

LysandreJik commented:
Merging this as soon as it's green, thank you for iterating on the PR! Sorry this took so long to merge.

enzoampil commented:
Thanks @LysandreJik and congrats @cceyda !! :smile:

Botfacke commented:
fix_bart_gpu

LysandreJik commented:
FYI this broke the behavior of the NER pipeline:

```py
from transformers import pipeline

nlp = pipeline("ner")

nlp("My name is Alex and I live in New York")
```

crashes with the following error:
```
    raise Exception("To decode [UNK] tokens use a fast tokenizer or provide offset_mapping parameter")
Exception: To decode [UNK] tokens use a fast tokenizer or provide offset_mapping parameter
```

Trying to see if this can be quickly patched, otherwise we'll revert the PR while we patch this.

cceyda commented:
oops! although returning unk tokens with slow tokenizers are not the best, I agree not forcing a fast tokenizer with a default of ignore_subword=True looks better for keeping the compatibility. I saw a bit late the _args_parser line was mis-merged during this pr merge and I see it is fixed/improved on the patch. I wasn't sure on how to test for the offset_mapping argument with the new test structure (which looks to be good at the patch). Sorry for the trouble 😅 @LysandreJik

LysandreJik commented:
No worries, thanks for taking a look at the patch!
